### PR TITLE
Expose rule stats in QueryStats

### LINKED ISSUE #2578: Expose rule stats in QueryStats
Currently we gather JMX stats for rule execution. However, we could also expose rule stats for particular query in `QueryStats`. This would simplify query debugging

### PR DESCRIPTION
**Pull request by Dith3r**
## Description

Fixes #2578.

Expose optimizer rule statistics per query in QueryInfo JSON. The number of rules exposed could be adjusted using the `query.reported-rule-stats-limit`  configuration parameter.

Example output:

```json
    "optimizerRulesSummaries" : [ {
    "rule" : "io.trino.sql.planner.iterative.rule.PushPredicateIntoTableScan",
    "invocations" : 10,
    "applications" : 1,
    "totalTime" : 999,
    "failures" : 0
  },
   (...)] 
```

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(x) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.airlift.stats.TDigest;
+import io.trino.plugin.base.metrics.TDigestHistogram;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class QueryRuleStats
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:24]: We already have `io.trino.sql.planner.iterative.RuleStats`. Why do we need another one?

Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:24]: `RuleStats` is not JSON serializable and it is used in a different context.

**Comment by Dith3r** [comment]
Failure due https://github.com/trinodb/trino/issues/16383

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollector.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                @Override
+                public List<QueryRuleStats> getQueryRulesStatsTopN()
+                {
+                    return emptyList();
+                }
+            };
+
+    void recordRule(Rule<?> rule, boolean invoked, boolean applied, long elapsedNanos);
+
+    <T> void recordRuleFailure(Rule<T> rule);
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollector.java:46]: do we need type parameter `T`?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java [hunk 3]
```diff
@@ -383,6 +396,7 @@ public void checkTimeoutNotExhausted()
 
         public void recordRuleInvocation(Rule<?> rule, boolean invoked, boolean applied, long elapsedNanos)
         {
+            queryStatsCollector.recordRule(rule, invoked, applied, elapsedNanos);
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java:399]: it seems that now we will have 3 rule stats collectors. Global, per query for timeout exhaustion and the new per query for query json. Do you think somehow merging those thres together makes sense? We could for example gather global stats through per query collector + use it to control the time.

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+{
+    private final String rule;
+    private long invocations;
+    private long hits;
+    private final TDigest time;
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:29]: do we need `time. digest` property? it takes a lot of space and it will grow query json that is too big already.

Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:29]: The number of rules to be presented in JSON is configurable, we don't present all rules statistics in response.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java (hunk 3 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java:399]: Merge two per query rule statistics.

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java (hunk 1 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:24]: One `RuleStats` class removed.

**Comment by Dith3r** [comment]
Failed due https://github.com/trinodb/trino/issues/16406

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 5]
```diff
@@ -72,6 +72,7 @@
     public static final String QUERY_MAX_EXECUTION_TIME = "query_max_execution_time";
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:72]: Add more description to commit message

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 6]
```diff
@@ -72,6 +72,7 @@
...
     public static final String QUERY_MAX_PLANNING_TIME = "query_max_planning_time";
     public static final String QUERY_MAX_RUN_TIME = "query_max_run_time";
+    public static final String QUERY_REPORTED_RULE_STATS_LIMIT = "query_reported_rule_stats_limit";
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:75]: Since query info is not a user visible change, therefore I don't think we should expose session property for `query_reported_rule_stats_limit`. Just feature config is enough

File: core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java [hunk 7]
```diff
@@ -406,6 +407,19 @@ public QueryManagerConfig setQueryMaxScanPhysicalBytes(DataSize queryMaxScanPhys
         return this;
     }
 
+    @Min(1)
+    public int getQueryReportedRuleStatsLimit()
+    {
+        return queryReportedRuleStatsLimit;
+    }
+
+    @Config("query.reported-rule-stats-limit")
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:416]: TBH, I would probably keep it as static

File: core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java [hunk 8]
```diff
@@ -201,6 +204,7 @@ private QueryStateMachine(
         this.queryStateTimer = new QueryStateTimer(ticker);
         this.metadata = requireNonNull(metadata, "metadata is null");
         this.stateMachineExecutor = requireNonNull(stateMachineExecutor, "stateMachineExecutor is null");
+        this.queryStatsCollector = queryStatsCollector;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java:207]: require not null

File: core/trino-main/src/main/java/io/trino/server/CoordinatorModule.java [hunk 9]
```diff
@@ -318,7 +320,6 @@ protected void setup(Binder binder)
         binder.bind(QueryExecutionMBean.class).in(Scopes.SINGLETON);
         newExporter(binder).export(QueryExecutionMBean.class)
                 .as(generator -> generator.generatedNameOf(QueryExecution.class));
-
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/server/CoordinatorModule.java:321]: undo

File: core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+import io.trino.sql.planner.iterative.Rule;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class DefaultQueryStatsCollector
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java:25]: why do we need multiple implementations?

Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java:25]: Similar design as with `WarningCollector` and its `NOOP` implementation.

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java (hunk 4 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:29]: `time` is serialized as `TDigestHistogram` as it could be read from JSON. `TimeDistribution` and `TimeDistributionSnapshot` would fail `TestQueryStats`

File: core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java (hunk 7 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:416]: I've followed your comment from the other PR. In some problematic cases, this parameter could easily give us more rule statistics than the default 20 without the need to deploy a new version of Trino. But it is up to you if you want to remove it.

File: core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java (hunk 10 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java:25]: When NOOP is used?

**Comment by sopel39** [comment]
Are the rules ordered by total time?

**Comment by Dith3r** [comment]
Yes, rules are ordered by `totalTime` in descending order.

File: core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java (hunk 10 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java:25]: Mostly in tests, but there are two usages in LocalQueryRunner.

**Comment by Dith3r** [comment]
Redshift test failing -> https://github.com/trinodb/trino/pull/16484

File: core/trino-main/src/main/java/io/trino/execution/QueryStats.java [hunk 11]
```diff
@@ -127,6 +128,7 @@
     private final DynamicFiltersStats dynamicFiltersStats;
 
     private final List<OperatorStats> operatorSummaries;
+    private final ImmutableList<QueryRuleStatsSnapshot> optimizerRulesSummaries;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/QueryStats.java:131]: `ImmutableList` -> `List`

File: core/trino-main/src/main/java/io/trino/execution/QueryStats.java [hunk 12]
```diff
@@ -771,6 +775,12 @@ public List<OperatorStats> getOperatorSummaries()
         return operatorSummaries;
     }
 
+    @JsonProperty
+    public ImmutableList<QueryRuleStatsSnapshot> getOptimizerRulesSummaries()
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/QueryStats.java:779]: `List`

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class QueryRuleStats
+{
+    private final String rule;
+    private long invocations;
+    private long applications;
+    private final TimeDistribution time = new TimeDistribution(TimeUnit.MICROSECONDS);
+    private long totalTime;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:32]: `time.getCount()` is the same (`totalTime`)

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    private long failures;
+
+    public QueryRuleStats(String rule)
+    {
+        this.rule = rule;
+    }
+
+    public void record(long nanos, boolean match)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:40]: nit: `match` -> `application`

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+        public QueryRuleStatsSnapshot(
+                @JsonProperty("rule") String rule,
+                @JsonProperty("invocations") long invocations,
+                @JsonProperty("applications") long applications,
+                @JsonProperty("time") TimeDistributionSnapshot time,
+                @JsonProperty("totalTime") long totalTime,
+                @JsonProperty("failures") long failures)
+        {
+            this.rule = requireNonNull(rule, "rule is null");
+            checkArgument(invocations >= 0, "invocations is negative");
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:112]: nit: I don't think it can happen, since it's internal class, we could maybe skip it

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollectorFactory.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.querystats;
+
+import io.trino.execution.QueryManagerConfig;
+
+import javax.inject.Inject;
+
+public class QueryStatsCollectorFactory
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollectorFactory.java:20]: Do we need this factory. IMO we could just create `DefaultQueryStatsCollector` in code. You already have `QueryManagerConfig` in `LocalDispatchQueryFactory`

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java [hunk 17]
```diff
@@ -349,22 +357,23 @@ public Context(
...
                 String message = format("The optimizer exhausted the time limit of %d ms", timeoutInMilliseconds);
-                if (ruleStats.isEmpty()) {
+                if (queryStatsCollector.isEmpty()) {
                     message += ": no rules invoked";
                 }
                 else {
                     long timeThresholdNanos = MILLISECONDS.toNanos(timeoutInMilliseconds) / 100;
-                    List<Entry<Rule<?>, RuleInvocationStats>> topRulesByTime = ruleStats.entrySet().stream()
+                    List<QueryRuleStats> topRulesByTime = queryStatsCollector.getQueryRulesStats().stream()
                             // Filter out rules that used less than 1% time that do not change the state
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java:374]: nit: I would probably just use `getSnapshot()`, or you could also have `getSnapshot(int limit)` as override.
Feel free to skip comment

File: docs/src/main/sphinx/admin/properties-query-management.rst [hunk 18]
```diff
@@ -132,6 +132,14 @@ After this period the coordinator will make its best effort to stop the
 query. Note that some operations in planning phase are not easily cancellable
 and may not terminate immediately.
 
+``query.reported-rule-stats-limit``
```
Comment (sopel39) [inline docs/src/main/sphinx/admin/properties-query-management.rst:135]: I would skip it. Query JSON is internal engine object, no need to expose details of it

**Review by sopel39 — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java (hunk 13 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:32]: from class `TimeDistribution`:
```
getAvg() {
        return this.convertToUnit(this.total.getCount()) / this.getCount();
    }

getCount() {
      return this.digest.getCount();
  }
```

So `getCount` returns the number of points added, `this.total.getCount()` as a sum of points is not exposed.

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollector.java [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+                {
+                    return emptyList();
+                }
+            };
+
+    void recordRule(Rule<?> rule, boolean invoked, boolean applied, long elapsedNanos);
+
+    void recordRuleFailure(Rule rule);
+
+    List<QueryRuleStats.QueryRuleStatsSnapshot> getQueryRulesStatsSnapshot();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollector.java:60]: static import QueryRuleStatsSnapshot

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        return totalTime;
+    }
+
+    public QueryRuleStatsSnapshot snapshot()
+    {
+        return new QueryRuleStatsSnapshot(rule, invocations, applications, time.snapshot(), totalTime, failures);
+    }
+
+    public static class QueryRuleStatsSnapshot
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:93]: QueryRuleStatsSnapshot -> PlanOptimizerStatistics

Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:93]: We can move this class to `io.trino.spi.eventlistener` and expose it to the event listener by adding the top-n list of these to the QueryStatistics class.

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.airlift.stats.TimeDistribution;
+import io.airlift.stats.TimeDistribution.TimeDistributionSnapshot;
+
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class QueryRuleStats
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java:26]: QueryRuleStats -> QueryPlanOptimizerStats

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollector.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+ */
+package io.trino.execution.querystats;
+
+import io.trino.sql.planner.iterative.Rule;
+
+import java.util.List;
+
+import static java.util.Collections.emptyList;
+
+public interface QueryStatsCollector
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollector.java:22]: Interface looks unnecessary to me.
You can add the below method to the implementation class and use that everywhere in tests.
```
public static QueryStatsCollector createTestingOptimizerStatsCollector()
{
     return new QueryStatsCollector(Integer.MAX_VALUE);
}
```

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollectorFactory.java (hunk 16 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryStatsCollectorFactory.java:20]: +1 looks unnecessary

File: core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/querystats/DefaultQueryStatsCollector.java:26]: DefaultQueryStatsCollector -> PlanOptimizersStatsCollector

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StatsRecordingPlanOptimizer.java [hunk 24]
```diff
@@ -45,13 +46,14 @@ public PlanNode optimize(
...
+            QueryStatsCollector queryStatsCollector,
             TableStatsProvider tableStatsProvider)
     {
         PlanNode result;
         long duration;
         try {
             long start = System.nanoTime();
-            result = delegate.optimize(plan, session, types, symbolAllocator, idAllocator, warningCollector, tableStatsProvider);
+            result = delegate.optimize(plan, session, types, symbolAllocator, idAllocator, warningCollector, queryStatsCollector, tableStatsProvider);
             duration = System.nanoTime() - start;
```
[the remark this answers is not in the document]
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StatsRecordingPlanOptimizer.java:57]: We don't measure optimizer time only, where internal rules statistics are taken.

[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StatsRecordingPlanOptimizer.java:57]: We need to measure the stats of rules derived from `PlanOptimizer` as well. Not all the planner rules have been migrated to the IterativeOptimizer framework.

**Comment by sopel39** [comment]
@Dith3r pine me when comments are addressed and tests pass

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java (hunk 20 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java]: SPI does not have `TimeDistributionSnapshot` class in dependencies.

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java (hunk 1 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java]: To clarify, class `RuleInvocationStats` was removed. Both classes (`RuleStats` and `QueryPlanOptimizerStats`) are used in different context, one is used for JMX reporting and second one to expose statistics for `QueryInfo` JSON for `Rule` and `PlanOptimizer`.

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQueryFactory.java [hunk 25]
```diff
@@ -97,7 +99,9 @@ public LocalDispatchQueryFactory(
         this.warningCollectorFactory = requireNonNull(warningCollectorFactory, "warningCollectorFactory is null");
         this.clusterSizeMonitor = requireNonNull(clusterSizeMonitor, "clusterSizeMonitor is null");
         this.executor = dispatchExecutor.getExecutor();
-        this.maxStateMachineThreadsPerQuery = requireNonNull(queryManagerConfig, "queryManagerConfig is null").getMaxStateMachineCallbackThreads();
+        requireNonNull(queryManagerConfig, "queryManagerConfig is null");
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQueryFactory.java:102]: This `requireNonNull` is unnecessary as queryManagerConfig is accessed on the next line. If it is a null we will get a similarly good error message from the next line (nullptr exception messages are improved by default from jdk 15 onwards)

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java [hunk 26]
```diff
@@ -349,87 +355,39 @@ public Context(
...
-
-            @Override
-            public String toString()
-            {
-                long elapsedMillis = getElapsedTimeMillis();
-                return format("%s ms, %s invocations, %s applications", elapsedMillis, invocationCount, applicationCount);
-            }
+            // record statistics per query
+            planOptimizersStatsCollector.recordRule(rule, invoked, applied, elapsedNanos);
+            // record statistics per optimizer
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java:389]: It will be better to move these code comments where these fields are defined for the class

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java (hunk 20 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java]: We can add a `TimeDistribution` to `io.trino.spi.eventlistener`, it's more convenient for the event listener to have visibility to the class used to serialize the json rather than a raw json string.
Feel free to do all the event listener changes as a follow up.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java [hunk 27]
```diff
@@ -349,87 +357,37 @@ public Context(
             this.timeoutInMilliseconds = timeoutInMilliseconds;
             this.session = session;
             this.warningCollector = warningCollector;
+            this.planOptimizersStatsCollector = planOptimizersStatsCollector;
+            this.iterativeOptimizerStatsCollector = new PlanOptimizersStatsCollector(5);
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java:361]: Having two collectors and recording stats twice is not great. We should figure out how to consolidate (by having the error reporting below use the provided collector instead, by moving the error reporting elsewhere, or by removing it altogether)

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java (hunk 27 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java:361]: Removed duplicate recording of statistics.

File: core/trino-main/src/main/java/io/trino/event/QueryMonitor.java [hunk 28]
```diff
@@ -288,7 +289,6 @@ private QueryStatistics createQueryStatistics(QueryInfo queryInfo)
         for (OperatorStats summary : operatorStats) {
             operatorSummaries.add(operatorStatsCodec.toJson(summary));
         }
-
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:291]: unrelated change

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java [hunk 29]
```diff
@@ -349,87 +355,40 @@ public Context(
             this.timeoutInMilliseconds = timeoutInMilliseconds;
             this.session = session;
             this.warningCollector = warningCollector;
+            this.iterativeOptimizerStatsCollector = new PlanOptimizersStatsCollector(5);
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java:358]: We shouldn't be using 5 limit here anymore as we want to merge everything back to the query level collector.
You can do the top 5 part in checkTimeoutNotExhausted, the limit could be a parameter to `getQueryRulesStatsSnapshot` as well.

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/QueryStatistics.java [hunk 30]
```diff
@@ -77,6 +77,11 @@
      * can change without preserving backward compatibility.
      */
     private final List<String> operatorSummaries;
+    /**
+     * Optimizer Rules summaries serialized to JSON. Serialization format and structure
+     * can change without preserving backward compatibility.
+     */
```
Comment (raunaqmorarka) [inline core/trino-spi/src/main/java/io/trino/spi/eventlistener/QueryStatistics.java:83]: This comment can be skipped now

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/QueryPlanOptimizerStatistics.java [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+ * limitations under the License.
+ */
+package io.trino.spi.eventlistener;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import static java.util.Objects.requireNonNull;
+
+// Serialization format and structure can change without preserving backward compatibility.
```
Comment (raunaqmorarka) [inline core/trino-spi/src/main/java/io/trino/spi/eventlistener/QueryPlanOptimizerStatistics.java:21]: nit: should be java doc like in the other classes in this package

**Review by raunaqmorarka — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java (hunk 27 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java]: @martint could you PTAL and drop the request changes if it's okay now ?

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/QueryPlanOptimizerStatistics.java [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+ */
+public record QueryPlanOptimizerStatistics(String rule, long invocations, long applications, long totalTime, long failures)
+{
+    @JsonCreator
+    public QueryPlanOptimizerStatistics(
+            @JsonProperty("rule") String rule,
+            @JsonProperty("invocations") long invocations,
+            @JsonProperty("applications") long applications,
+            @JsonProperty("totalTime") long totalTime,
+            @JsonProperty("failures") long failures)
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/eventlistener/QueryPlanOptimizerStatistics.java:32]: You should be able to annotate the record declaration directly instead of needing an explicit constructor.

File: core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java [hunk 33]
```diff
@@ -0,0 +1,81 @@
...
+                .record(duration, true);
+    }
+
+    public void recordFailure(Object rule)
+    {
+        stats.computeIfAbsent(rule.getClass(), (key) -> new QueryPlanOptimizerStats(key.getCanonicalName()))
+                .recordFailure();
+    }
+
+    public List<QueryPlanOptimizerStatistics> getTopNQueryRulesStats()
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:57]: Rename to getTopRuleStats

File: core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java [hunk 34]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        return getTopNQueryRulesStats(queryReportedRuleStatsLimit);
+    }
+
+    public List<QueryPlanOptimizerStatistics> getTopNQueryRulesStats(int limit)
+    {
+        return stats.entrySet().stream()
+                .sorted(Comparator.<Map.Entry<Class<?>, QueryPlanOptimizerStats>, Long>comparing(queryPlanOptimizerStatsEntry -> queryPlanOptimizerStatsEntry.getValue().getTotalTime()).reversed())
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:65]: Rename `queryPlanOptimizerStatsEntry` to `entry` for conciseness.

File: core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java [hunk 35]
```diff
@@ -0,0 +1,81 @@
...
+                .limit(limit)
+                .map((Map.Entry<Class<?>, QueryPlanOptimizerStats> queryPlanOptimizerStatsEntry) -> queryPlanOptimizerStatsEntry.getValue().snapshot(queryPlanOptimizerStatsEntry.getKey().getCanonicalName()))
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:67]: Rename `queryPlanOptimizerStatsEntry` to `entry` for conciseness.

File: core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java [hunk 36]
```diff
@@ -0,0 +1,81 @@
...
+                .collect(toImmutableList());
+    }
+
+    public void add(PlanOptimizersStatsCollector optimizerStatsCollector)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:71]: Rename to `collector` for conciseness

File: core/trino-main/src/main/java/io/trino/execution/querystats/QueryPlanOptimizerStats.java [hunk 37]
```diff
@@ -0,0 +1,85 @@
...
+    private long applications;
+    private long totalTime;
+    private long failures;
+
+    public QueryPlanOptimizerStats(String rule)
+    {
+        this.rule = rule;
+    }
+
+    public void record(long nanos, boolean application)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/querystats/QueryPlanOptimizerStats.java:31]: Renamed `application` to `applied`

File: core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java [hunk 38]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    public void recordOptimizer(PlanOptimizer planOptimizer, long duration)
+    {
+        stats.computeIfAbsent(planOptimizer.getClass(), (key) -> new QueryPlanOptimizerStats(key.getCanonicalName()))
+                .record(duration, true);
+    }
+
+    public void recordFailure(Object rule)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:51]: Sorry, I missed this one earlier. Why is this `Object` and not `Rule<?>`. If the reason is because of `PlanOptimizer` vs `Rule` being unrelated, then let's introduce two overloads to make the code more type-safe.

**Comment by colebow** [comment]
Because this includes a new configuration property, it needs documentation. Would you be able to do a follow-up PR for that, or would you prefer that I handle it?

cc @martint

**Comment by sopel39** [comment]
cc @Dith3r ^^

**Comment by Dith3r** [comment]
In one of the patchsets I've removed documentation as we agreed that this is very internal thing (query JSON is only affected). If we want to add documentation, I can create new PR for it - please only confirm if this is needed.

**Comment by findepi** [comment]
follow-up -- https://github.com/trinodb/trino/issues/16960
